lib/util.h

```diff
@@ -289,13 +289,13 @@ struct buf {
 #define buf_ensure(b, n) do { if ((b)->alloc < (b)->len + (n)) _buf_ensure((b), (n)); } while (0)
 #define buf_putc(b, c) do { buf_ensure((b), 1); (b)->s[(b)->len++] = (c); } while (0)
 
-void _buf_ensure(struct buf *buf, size_t len);
-const char *buf_cstring(const struct buf *buf);
-const char *buf_cstringnull(const struct buf *buf);
-const char *buf_cstringnull_ifempty(const struct buf *buf);
-char *buf_release(struct buf *buf);
-char *buf_newcstring(struct buf *buf);
-char *buf_releasenull(struct buf *buf);
+__attribute__((nonnull)) void _buf_ensure(struct buf *buf, size_t len);
```

Comment:
I'm all for helping the compiler but it seems arbitrary why to just annotate these functions. And if we want annotate all headers, then we should discuss this with the rest of the Cyrus team. The cyrus-devel list is a better place to agree on this.

Reply:
The initial problem was, that the Clang Static Analyzer complained in one of the `desc ? desc : x` checks, that in case desc is NULL, later in the code bad things will happen (I do not remember which), and looking further it turend out, that desc cannot be NULL.  Hence I added annotations to all involved functions.

Reply:
After more thought I come to agree and my initial comment is moot. Setting the attributes to help the compiler for these functions is definitely going to be useful, thanks. The point that @elliefm raised in https://github.com/cyrusimap/cyrus-imapd/pull/4062#discussion_r862542301 still stands.

Reply:
> The point that @elliefm raised in [#4062 (comment)](https://github.com/cyrusimap/cyrus-imapd/pull/4062#discussion_r862542301) still stands.

The point I raised in https://github.com/cyrusimap/cyrus-imapd/pull/4062#discussion_r868007376 still stands.

Reply:
This is not how we do code reviews in this project. It is great that your patch highlights an issue in the current error handling, but your proposed code change does not actually fix the issue. In such a situation, we typically either include the fix right with the pull request (as a separate commit), or we keep the current erroneous code section and create a new pull request for the fix.

I suggest you either remove the disputed code from your PR, or you change the code to include the actual fix. It is very rare that we have two authors on one pull request and I do not see a reason to change that for this one.

Reply:
In this project it is random, whether code review will be done.

I meant you change the code as you think it shall be changed.  Once that changes are upstream, I will adapt the changes here.

Besides, the number of authors of a pull requests is irrelevant.

Reply:
I just landed https://github.com/cyrusimap/cyrus-imapd/commit/f5258901cb66e4b0084da9035309630ff26ca17d upstream